GH-319: [Java] recipe to concatenate value vectors as one (#320)
apache/arrow-cookbook: java/source/data.rst

.. contents::

Concatenate Value Vectors
=========================

In some cases, we need to concatenate two value vectors into one. To accomplish

> this, we can use `VectorAppender`_. Consider also that the initial Value Vector
> is mutated and used to concatenate the added values.

Reviewer: ```suggestion
this, we can use `VectorAppender`_. This mutates the initial ValueVector.
```

.. testcode::

    import org.apache.arrow.memory.BufferAllocator;
    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.vector.IntVector;
    import org.apache.arrow.vector.ValueVector;
    import org.apache.arrow.vector.util.VectorAppender;

    try (
        BufferAllocator allocator = new RootAllocator();
        IntVector initialValues = new IntVector("initialValues", allocator);
        IntVector toAppend = new IntVector("toAppend", allocator);
    ) {
      initialValues.allocateNew(2);
      initialValues.set(0, 1);
      initialValues.set(1, 2);
      initialValues.setValueCount(2);
      System.out.println("Initial IntVector: " + initialValues);
      toAppend.allocateNew(4);
      toAppend.set(1, 4);
      toAppend.set(3, 6);
      toAppend.setValueCount(4);
      System.out.println("IntVector to Append: " + toAppend);
      VectorAppender appenderUtil = new VectorAppender(initialValues);

>       ValueVector resultOfVectorsAppended = toAppend.accept(appenderUtil, null);

Reviewer: We don't need to close this? If it mutates the original vector, we should write it in a way that makes that clear

Author: Do you have any ideas why the leak wasn't caught? Maybe this ValueVector is also deleted after the dependencies were closed?

Reviewer: probably because it mutates the original vector instead of allocating a new one...

Author: Added more detail

Reviewer: Just...don't use the return value.

Author: I have updated the recipe to not use the return value. It is necessary to analyze the change upstream to not return value since it is part of the Value Vector.
